### PR TITLE
Ui tests

### PR DESCRIPTION
**Pull request by JasonStoltz**
Pushing up some more unit tests.

Instead of starting at the top and working down like last time, I'm starting at the bottom and working up.

There's more to come, but just wanted to get some out there.

### CODE REVIEW

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { mount } from 'enzyme';
```
Comment (JasonStoltz) [inline x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: I use mount over shallow in most tests for two reasons...

1. Since we use EUI components it would require a lot of diving, which starts to make these tests a bit more brittle, because any UI tweak would also require cause us to adjust our dive chain. Using mount I just mount the whole thing and pick out interesting elements to test using `data-test-subj`.

2. In order to execute `useContext`, we need to use `mount`, afaict.

I this also generally lets us mock less and test bigger unit chunks.

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        expect(subject().props().isDisabled).toBeFalsy();
+      });
+
+      it('should use the enterpriseSearchUrl as the base path for its href', () => {
+        expect(subject().props().href).toBe('http://localhost:3002/as');
+      });
+    });
+  });
+
+  describe('when enterpriseSearchUrl is not set', () => {
```
Comment (richkuz) [inline x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx:35]: From an end user perspective, what circumstances would cause KibanaContext to not provide our plugin with an enterpriseSearchUrl? I'm trying to understand better if disabling the Launch App Search button is the appropriate behavior here.

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    const eventMock = {
+      preventDefault: jest.fn(),
+    };
+
+    it('has a link to Enterprise Search Home page first', () => {
+      subject()[0].onClick(eventMock);
+      expect(historyMock.push).toHaveBeenCalledWith('/');
+    });
+
+    it('has a link to our custom page third', () => {
```
Comment (richkuz) [inline x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts]: replace with: `it('has a link to page 1 second')`

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      expect(appSearchBreadCrumbsInnerCall).toHaveBeenCalledWith([
+        { text: 'Page 1', path: '/current-path' },
+      ]);
+
+      // then registers them with Kibana
+      expect(setBreadcrumbs).toHaveBeenCalledWith(builtBreadcrumbs);
+    });
+  });
+
+  describe('when isRoot is true', () => {
```
Comment (richkuz) [inline x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx:63]: I like the unit tests in this file. From reading the tests alone, I get a sense of what `<SetSearchBreadCrumbs/>` is intended to do: take in some breadcrumbs, and set them in Kibana. It makes it easy to test the behavior of the component without having to load all of Kibana. It verifies that the external touchpoints (e.g. Kibana's `setBreadCrumbs` API) are called.

I'm not exactly sure _why_ when `isRoot` is true we build breadcrumbs with an empty breadcrumb, but I imagine the docs for the component itself would explain why this is a useful and necessary thing to do.

File: x-pack/plugins/enterprise_search/public/applications/shared/react_router_helpers/link_events.test.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { letBrowserHandleEvent } from '../react_router_helpers';
```
Comment (richkuz) [inline x-pack/plugins/enterprise_search/public/applications/shared/react_router_helpers/link_events.test.ts:7]: I imagine if I were TDD'ing `letBrowserHandleEvent`, the unit test you wrote here is the kind of unit test I would write while developing the method. This unit test would help me test the code without having to load everything up in the browser. This would result in faster dev/test loops while building out the method. After the `letBrowserHandleEvent` method is implemented, and these unit tests were passing, I imagine I would then verify it manually in the browser.
👍

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    },
+  ];
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  const subject = () => appSearchBreadcrumbs(historyMock)(breadCrumbs);
+
+  it('Builds a chain of breadcrumbs with Enterprise Search and App Search at the root', () => {
```
Comment (richkuz) [inline x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts:36]: The tests in this file feel a bit like testing the presentation, like a thinly veiled snapshot test. I question if this test would have a long shelf life. It also doesn't test any edge cases. Maybe a higher-level jest integration test, one that actually tests `letBrowserHandleEvent`, but still one that mocks Kibana, and interacts with the page, would provide more value. 🤷 
Was this test helpful in TDD'ing `appSearchBreadcrumbs`?

**Review by richkuz — COMMENTED** [review]
None of my feedback is blocking, just adding to the discussion.

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 2 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx:35]: enterpriseSearchUrl is a configuration provided in a config file, by the user.

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts (hunk 6 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts:36]: This test will do 2 things:

- Documents what exactly the `appSearchBreadcrumbs` method does. It is not apparent by looking at the code, and it wasn't until I wrote this test that I understood what exactly it does. So it will serve as a helpful reference if anyone needs to go change this code at some point.
- Additionally, if someone ever needs to go modify this code or refactor it, these tests will help them to be confident that their changes are working as expected.

I personally like this test, it gives me confidence about what this unit does and that it is working as expected.

Could you elaborate on what would give this a short shelf life?

As to your suggestion about a higher level jest test, what would that look like? Would that just be this test with `letBrowserHandleEvent` un-mocked? What else would need to change?

File: x-pack/plugins/enterprise_search/public/applications/shared/react_router_helpers/link_events.test.ts (hunk 5 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/shared/react_router_helpers/link_events.test.ts:7]: It can also help you find cases you've missed. I made a small update to `link_events` as a result of writing this test.

Comment (cee-chen) [reply x-pack/plugins/enterprise_search/public/applications/shared/react_router_helpers/link_events.test.ts:7]: Just wanted to add here that these are super awesome/thorough tests! 🎉

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 1 above)
Comment (cee-chen) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: My wariness of using `mount()` over `shallow()` stems primarily from my past experiences writing hundreds of Jest tests for a large UI library. The primary dev impact I recall is one of speed - Jest tests are simply much faster (for obvious reasons) with `shallow()`. This only comes into play when we have a larger codebase with a lot of tests, which we don't have now - but we eventually will once AS moves fully over.

That being said, I accept that is a pre-optimization we don't have to worry about yet, that can easily wait until post-MVP, and you're right that useContext is far more annoying to do in `shallow()` vs. `mount()` (it's possible from my quick research, but it requires some extra setup that we're not doing).

I am slightly more concerned about point 1 - again, this is from past experience with writing tests with 3rd party libs (or even 1st party deps that we'd written ourselves) - but I tend to find that `mount()` tempts people into the trap of diving into a component that doesn't belong to us and that we shouldn't depend upon for tests. From a high level point of view for a unit test, all we care about is that a component named "EuiButton" rendered, and that the correct props were passed to it - not anything that's inside it (that belongs to the unit test for EuiButton itself, which EUI is in charge of).

For example, if we wrote something like `wrapper.find('[data-test-subj="someStringSetByEui"]'` - that's now fragile and causes failures if a 3rd party lib changes without warning. We didn't do so in this test, which is great, but I find that `shallow()` generally tends to discourage / makes it much harder to write code that relies on 3rd party deps by forcing devs to manually dive into components rather than just having everything readily available.

Sorry, all that was a lot of words :) TL;DR: I'm definitely not against moving forward with `mount()`, but especially with performance considerations down the road, I'd like us to be open to moving to `shallow()` as a tech debt item at some point. Does that sound good?

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts (hunk 6 above)
Comment (richkuz) [reply x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts:36]: > Documents what exactly the appSearchBreadcrumbs method does

👍  "Tests as documentation for what the code does" -- that's another value we should pin somewhere in our eventual Testing Principles Guide.

I went back and looked at the corresponding `appSearchBreadcrumbs` implementation. Now, seeing how it is written, I also like this test. I see that the test passes additional breadcrumb configuration into the method, and this test makes it clear the order in which operations are intended to be applied.

I (mistakenly) originally thought this was testing only the hard-coded configuration of the breadcrumbs (e.g. `/` comes first, `/app_search` comes next), which I assumed would be brittle. The test serves a deeper purpose than that. Comment redacted :)

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx (hunk 4 above)
Comment (richkuz) [reply x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx:63]: And... it's clear to me now after having read the implementation of `SetAppSearchBreadcrumbs` why `isRoot` works the way it does. Cool.

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 2 above)
Comment (richkuz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx:35]: 👍

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    );
+  };
+
+  describe('when enterpriseSearchUrl is set', () => {
+    let wrapper;
+
+    beforeEach(() => {
+      enterpriseSearchUrl = 'http://localhost:3002';
+      wrapper = render();
+    });
```
Comment (cee-chen) [inline x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: I'd like to see us refactor everything within context to be a passed param rather than let strings.

As an example:

```jsx
  const render = context => {
    return mount(
      <KibanaContext.Provider
        value={{
          http: {},
          enterpriseSearchUrl: '',
          setBreadcrumbs: jest.fn(),
          ...context,
        }}
      >
        <EngineOverviewHeader />
      </KibanaContext.Provider>
    );
  };


  describe('when enterpriseSearchUrl is set', () => {
    beforeEach(() => {
      wrapper = render({ enterpriseSearchUrl: 'http://localhost:3002' });
    });
  });

  describe('when enterpriseSearchUrl is not set', () => {
    beforeEach(() => {
      wrapper = render({ enterpriseSearchUrl: undefined });
    });
  });
```

Let me know how that feels to you?

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    describe('the Launch App Search button', () => {
+      const subject = () => wrapper.find('EuiButton[data-test-subj="launch_button"]');
```
Comment (cee-chen) [inline x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: This is a very minor thing, but I would actually slightly prefer that we remain as library-agnostic as possible and instead just do `wrapper.find('[data-test-subj="launch_button"]')`. That way in the very unlikely scenario where we move away from EUI (or EUI changes their names?), we'll have less static strings to change, but primarily it's a best practice thing and one we also use in our E2E/Cypress tests.

As an alternative note - `snake_casing` is primarily for file/folder names in Kibana I believe, I believe we should be using `camelCase` for data attrs

File: x-pack/plugins/enterprise_search/public/application.test.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+/*
```
Comment (cee-chen) [inline x-pack/plugins/enterprise_search/public/application.test.ts]: This should be renamed `index.test.ts`

Comment (cee-chen) [reply x-pack/plugins/enterprise_search/public/application.test.ts]: Ah, and should also be moved to `public/applications/index.test.ts`

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        value={{
+          http: {},
+          enterpriseSearchUrl,
+          setBreadcrumbs: jest.fn(),
+        }}
+      >
+        <EngineOverviewHeader />
+      </KibanaContext.Provider>
+    );
+  };
```
Comment (cee-chen) [inline x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: I'd like to pull this out to a reusable helper function if possible. Something like this:

```jsx
const mountWithContext = (component, context) => {
  const defaultContext = {
    http: jest.fn(),
    setBreadcrumbs: jest.fn(),
    enterpriseSearchUrl: 'http://localhost:3002',
  };

  return mount(
    <KibanaContext.Provider value={{ ...defaultContext, ...context }}>
      {component}
    </KibanaContext.Provider>
  );
};
```
example usage:

```jsx
const wrapper = mountWithContext(<EngineOverviewHeader />, { enterpriseSearchUrl: '' });
```

I'd strongly prefer `mountWithContext()` over `render()` - [`render()` is already an existing OOTB function within Enzyme](https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/render.html) and one that most people find confusing vs shallow/mount and that I personally prefer to avoid.

My thought is that this could get exported from `index.test.ts`, but maybe `helpers.test.js` is better? 🤷 I'd normally put in a Jest [setupFile](https://jestjs.io/docs/en/configuration.html#setupfiles-array) but unfortunately since we're inheriting Jest settings from Kibana, I don't think we can make this a global helper without an import.

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 7 above)
Comment (cee-chen) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: Final form here: https://github.com/constancecchen/kibana/pull/4/files#r415998493

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 8 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: `wrapper.find('[data-test-subj="launch_button"]')` actually finds two elements for some reason, I haven't worked out exactly why, because there is only one rendered element in the `.html()` output. `'EuiButton[data-test-subj="launch_button"]'` was the only way I could properly select the element I was looking for. So this was basically a shortcut because I didn't feel like going any deeper into this problem right now. Happy to take a second look though.

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 10 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: I like it.

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 1 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: I 100% agree with you. I usually default to shallow, and if shallow won't work, I use mount.

One way I've handled the context thing in components is export two versions of a component, the outer component, which wraps an inner component to access context and pass it through as a simple property to the inner component. Then I unit test the InnerComponent.

```
export default OuterComponent = () => {
  const {prop} = useContext(...)
  return <InnerComponent prop={prop}
}

export const InnerComponent = ({prop}) => {
  <div>Hi, I am a {prop}</div>
}
```

Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: That was pre-hooks, redux higher-order component days, things are probably different now though.

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 8 above)
Comment (cee-chen) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: Interesting, I'm seeing the same thing. I think it's because EuiButton is passing `data-test-subj` down to as a prop to the raw `<button>` component, but the way Enzyme works, the prop is available on both the parent EuiButton component and the button child.

I'm fairly certain this wouldn't be an issue with shallow, so I'm good with leaving it for now and come back to it later when we've investigated using shallow more thoroughly.

If we could change to camelCasing though that would be dope 👍

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx (hunk 4 above)
Comment (cee-chen) [reply x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx:63]: Sorry, I think this was bad prop naming on my part - I kind of hand-waved a flag that I thought sounded good for "When the breadcrumb is just the 'App Search' base") but I think that could be cleaned up significantly.

My original goal was to make it easy to do nested breadcrumbs, like so

```jsx
appSearchBreadcrumbs(history)([
  { text: 'foo', path: '/foo' },
  { text: 'bar', path: '/foo/bar' },
]);
```

but I'm not totally sure I succeeded since it made setting a base App Search path (aka `isRoot`) with no child breadcrumbs a little opaque. 😬 

I'm definitely open to rewriting the source code here!

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 10 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: I created a `test_utils/helpers.tsx` file, which seemed consistent with what Kibana does with other test helpers like their enzyme helper. a7b0acd579ac21f46e165ecc49ec1a0643b10d1f

File: x-pack/plugins/enterprise_search/public/application.test.ts (hunk 9 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/application.test.ts]: a7b0acd

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 7 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: Feels good a7b0acd

File: x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx (hunk 8 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx]: Dope. https://github.com/constancecchen/kibana/pull/4/commits/5f28efe392a5b59aa1fd64cd537183d5687f905c

**Comment by JasonStoltz** [comment]
@constancecchen Makes sense, updated here: https://github.com/constancecchen/kibana/pull/4/commits/8a678a0e2fb1218fee431d6989bf65a03bbd494c.

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts (hunk 3 above)
Comment (JasonStoltz) [reply x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/generate_breadcrumbs.test.ts]: Updated here: https://github.com/constancecchen/kibana/pull/4/commits/a7b0acd579ac21f46e165ecc49ec1a0643b10d1f

File: x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  useHistory: () => ({
+    createHref: jest.fn(),
+    push: jest.fn(),
+    location: {
+      pathname: '/current-path',
+    },
+  }),
+}));
+
+import { KibanaContext } from '../..';
```
Comment (cee-chen) [inline x-pack/plugins/enterprise_search/public/applications/shared/kibana_breadcrumbs/set_breadcrumbs.test.tsx]: I think you can lose this line now that you're using `mountwithKibanaContext`, right?
```suggestion

```

**Review by cee-chen — APPROVED** [review]
🎉
